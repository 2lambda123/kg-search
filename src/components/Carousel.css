/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

.kgs-carousel {
    z-index: 1200;
    display: block;
    position: fixed;
    top: 0;  
    left: 0;  
    width: 100%;
    height: 100%;
    margin: 0;
    box-sizing: border-box;
    box-sizing: border-box;
    background-color: rgba(0,0,0,0.5);
}

.kgs-carousel__panel {
    display: block;
    position: relative;
    width: calc(100% - 20px);
    height: calc(100% - 20px);
    margin-top: 10px;
    margin-left: 10px;
    border-radius: 3px;
}

@media (min-width: 992px) {
    .kgs-carousel__panel {
        width: 992px;
        margin-left: calc(50% - 496px);
    }
}

@media (min-width: 1400px) {
    .kgs-carousel__panel {
        width: 1200px;
        margin-left: calc(50% - 600px);
    }
}

.kgs-carousel__item {
    z-index: 1200;
    display: block;
    position: absolute;
    top: 0;
    left: 200%;
    width: 100%;
    height: 100%;
    border-radius: 3px;
    background-color: white;
    transition: left 0.25s ease-in;
    transform: translate3D(0,0,0);
    background-color: var(--border-color-3);
}

.kgs-carousel .kgs-carousel__item.position0 {
    left: 0;
    display: block;
}

.kgs-carousel .kgs-carousel__item.position1 {
    left: 105%;
    display: block;
}

.kgs-carousel .kgs-carousel__item.position2 {
    left: 105%;
    display: none;
}

.kgs-carousel .kgs-carousel__item.position3 {
    left: -105%;
    display: none;
}

.kgs-carousel .kgs-carousel__item.position4 {
    left: -105%;
    display: block;
}

@media (min-width: 992px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 205%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -205%;
    }
}

@media (min-width: 1984px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 305%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -305%;
    }
}

@media (min-width: 2976px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 405%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -405%;
    }
}

@media (min-width: 3968px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 505%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -505%;
    }
}

@media (min-width: 4960px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 605%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -605%;
    }
}

.kgs-carousel__content {
    position: relative;
    width: 100%;
    height: 100%;
    background-color: var(--background-color-2);
    border-radius: 3px;
}

.kgs-carousel__body {
    position: relative;
    width: 100%;
    height: calc(100% - 50px);
    border-top: 1px solid var(--border-color-3);
    overflow-y: auto;
}

/*
.kgs-carousel .kgs-carousel__item.position0 .kgs-carousel__body {
    background-color: red;
}

.kgs-carousel .kgs-carousel__item.position1 .kgs-carousel__body {
    background-color: green;
}

.kgs-carousel .kgs-carousel__item.position2 .kgs-carousel__body {
    background-color: blue;
}

.kgs-carousel .kgs-carousel__item.position3 .kgs-carousel__body {
    background-color: yellow;
}

.kgs-carousel .kgs-carousel__item.position4 .kgs-carousel__body {
    background-color: maroon;
}
*/

.kgs-carousel__header {
    display: flex;
    position: relative;
    width: 100%;
    height: 50px;
    box-sizing: border-box;    
    /*
    box-shadow: 0 0 4px rgba(0,0,0,.14), 0 4px 8px rgba(0,0,0,.28);
    */
}

@media screen and (min-width:0\0) {  
    /* IE9 and IE10 rule sets go here */  
    .kgs-carousel__header {
        top: 10px;
    }
}

@media (min-width: 992px) {
    .kgs-carousel__header {
        width: 992px;
    }
}

@media (min-width: 1400px) {
    .kgs-carousel__header {
        width: 1200px;
    }
}

.kgs-carousel__previous-button {
    height: 40px;
    margin: 5px 0 0 5px;
    padding: 7px 18px 7px 8px;
    background-color: var(--background-color-21);
    border: 0;
    border-radius: 3px;
    color: var(--color-21);
    font-weight: 500;
    transition: color ease-in-out .15s, box-shadow ease-in-out .15s;
}

/* small hack to fix adjustment for ie 11 */
@media all and (-ms-high-contrast:none) {
    .kgs-carousel__close-button { margin-top: -5px; } /* IE10 */
    *::-ms-backdrop, .kgs-carousel__close-button { margin-top: -5px; } /* IE11 */
}

.kgs-carousel__previous-button:hover {
    box-shadow: 3px 3px 6px var(--box-shadow-2);
    color: var(--color-22);
}

.kgs-carousel__navigation {
    flex: 1;
    margin: 0 10px;
    text-align: right;
}

.kgs-carousel__close-button {
    width: 40px;
    height: 40px;
    margin: 5px 7px 0 0;
    padding: 1px 7px 2px;
    border: 0;
    background-color: transparent;
    color: var(--color-19);
    text-align: center;
    transition: box-shadow 0.2s ease-in, color 0.2s ease-in;
    z-index: 1500;
}

/* small hack to fix adjustment for ie 11 */
@media all and (-ms-high-contrast:none) {
    .kgs-carousel__close-button { margin-top: -8px; } /* IE10 */
    *::-ms-backdrop, .kgs-carousel__close-button { margin-top: -8px; } /* IE11 */
}

.kgs-carousel__close-button:hover {
    box-shadow: 3px 3px 6px var(--box-shadow-2);
    color: var(--color-20);
}


.kgs-carousel__close-button::after {
    content: "esc";
    position: absolute;
    transform: translate(-16px,14px);
    font-size: var(--font-size-smaller);
    color: var(--color-8);
}