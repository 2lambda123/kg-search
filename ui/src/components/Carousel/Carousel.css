/*
 * Copyright 2018 - 2021 Swiss Federal Institute of Technology Lausanne (EPFL)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This open source software code was developed in part or in whole in the
 * Human Brain Project, funded from the European Union's Horizon 2020
 * Framework Programme for Research and Innovation under
 * Specific Grant Agreements No. 720270, No. 785907, and No. 945539
 * (Human Brain Project SGA1, SGA2 and SGA3).
 *
 */

.kgs-carousel {
    z-index: 1200;
    display: block;
    position: fixed;
    top: 0;  
    left: 0;  
    width: 100%;
    height: 100%;
    margin: 0;
    box-sizing: border-box;
    background-color: rgba(0,0,0,0.5);
}

.kgs-carousel__panel {
    display: block;
    position: relative;
    width: calc(100% - 20px);
    height: calc(100% - 20px);
    margin-top: 10px;
    margin-left: 10px;
    border-radius: 3px;
}

@media (min-width: 992px) {
    .kgs-carousel__panel {
        width: 992px;
        margin-left: calc(50% - 496px);
    }
}

@media (min-width: 1400px) {
    .kgs-carousel__panel {
        width: 1200px;
        margin-left: calc(50% - 600px);
    }
}

.kgs-carousel__item {
    z-index: 1200;
    display: block;
    position: absolute;
    top: 0;
    left: 200%;
    width: 100%;
    height: 100%;
    border-radius: 3px;
    transition: left 0.25s ease-in;
    transform: translate3D(0,0,0);
    background-color: var(--border-color-3);
}

.kgs-carousel .kgs-carousel__item.position0 {
    left: 0;
    display: block;
}

.kgs-carousel .kgs-carousel__item.position1 {
    left: 105%;
    display: block;
}

.kgs-carousel .kgs-carousel__item.position2 {
    left: 105%;
    display: none;
}

.kgs-carousel .kgs-carousel__item.position3 {
    left: -105%;
    display: none;
}

.kgs-carousel .kgs-carousel__item.position4 {
    left: -105%;
    display: block;
}

@media (min-width: 992px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 205%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -205%;
    }
}

@media (min-width: 1984px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 305%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -305%;
    }
}

@media (min-width: 2976px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 405%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -405%;
    }
}

@media (min-width: 3968px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 505%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -505%;
    }
}

@media (min-width: 4960px) {
    .kgs-carousel .kgs-carousel__item.position1,
    .kgs-carousel .kgs-carousel__item.position2 {
        left: 605%;
    }
    
    .kgs-carousel .kgs-carousel__item.position3,
    .kgs-carousel .kgs-carousel__item.position4 {
        left: -605%;
    }
}

.kgs-carousel__content {
    position: relative;
    width: 100%;
    height: 100%;
    background-color: var(--background-color-2);
    border-radius: 3px;
    overflow-y: auto;
}

/*
.kgs-carousel .kgs-carousel__item.position0 .kgs-carousel__body {
    background-color: red;
}

.kgs-carousel .kgs-carousel__item.position1 .kgs-carousel__body {
    background-color: green;
}

.kgs-carousel .kgs-carousel__item.position2 .kgs-carousel__body {
    background-color: blue;
}

.kgs-carousel .kgs-carousel__item.position3 .kgs-carousel__body {
    background-color: yellow;
}

.kgs-carousel .kgs-carousel__item.position4 .kgs-carousel__body {
    background-color: maroon;
}
*/

.kgs-carousel__header {
    display: flex;
    position: relative;
    width: 100%;
    height: 50px;
    box-sizing: border-box; 
    padding: 0 0 0 10px;   
    /*
    box-shadow: 0 0 4px rgba(0,0,0,.14), 0 4px 8px rgba(0,0,0,.28);
    */
}

@media screen and (min-width:0\0) {  
    /* IE9 and IE10 rule sets go here */  
    .kgs-carousel__header {
        top: 10px;
    }
}

@media (min-width: 900px) {

    .kgs-carousel__header {
        padding: 0 0 0 20px; 
    }
    
}

.kgs-carousel__previous-button {
    height: 40px;
    margin: 5px 0 0 5px;
    padding: 7px 18px 7px 8px;
    background-color: transparent;
    border: 0;
    border-radius: 3px;
    color: var(--color-2);
    font-weight: 500;
}

.kgs-carousel__previous-button:hover {
    font-weight: bold;
}

.kgs-carousel__previous-button:hover .fa-chevron-left:before {
    font-weight: bold;
}

.kgs-carousel__navigation {
    flex: 1;
    text-align: right;
}

.kgs-carousel__close-button {
    width: 40px;
    height: 40px;
    margin: 5px 7px 0 0;
    padding: 1px 7px 2px;
    border: 0;
    background-color: transparent;
    color: var(--color-2);
    text-align: center;
    font-weight: 500;
    z-index: 1500;
}

.kgs-carousel__close-button .fa-close:before {
    font-weight: 500;
}


.kgs-carousel__close-button:hover {
    font-weight: bold;
}

.kgs-carousel__close-button:hover .fa-close:before {
    font-weight: bold;
}

.kgs-carousel__close-button::after {
    content: "esc";
    position: absolute;
    transform: translate(-16px,14px);
    font-size: var(--font-size-extra-small);
    color: var(--color-2);
}

.kgs-carousel .kgs-info .kgs-info__panel {
    top: 0px;
    right: 0;
    bottom: 0px;
    left: 0px;
}

@media (min-width: 992px) {
    .kgs-carousel .kgs-info .kgs-info__panel  {
        margin-left: calc(50% - 496px);
    }
}
