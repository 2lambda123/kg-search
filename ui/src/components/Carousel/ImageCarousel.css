/*
 * Copyright 2018 - 2021 Swiss Federal Institute of Technology Lausanne (EPFL)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This open source software code was developed in part or in whole in the
 * Human Brain Project, funded from the European Union's Horizon 2020
 * Framework Programme for Research and Innovation under
 * Specific Grant Agreements No. 720270, No. 785907, and No. 945539
 * (Human Brain Project SGA1, SGA2 and SGA3).
 *
 */

.kgs-image_carousel > div .carousel .thumbs-wrapper {
  margin-top: 15px;
  margin-bottom: 0;
}

.kgs-image_carousel > div .carousel .thumbs-wrapper ul.thumbs {
  margin-bottom: 0;
}

.kgs-image_carousel .carousel .slide  {
    cursor: pointer;
}
  
.kgs-image_carousel .carousel .slide img  {
    min-height: 300px;
}
  
.kgs-image_carousel .carousel .thumb {
    border-color: var(--background-color-2);
    min-height: 80px;
}

.kgs-image_carousel .carousel .thumb.selected, 
.kgs-image_carousel .carousel .thumb:hover {
    border-color: var(--border-color-5);
}

.kgs-image_carousel .carousel .slide .legend {
  bottom: 0; 
  width: auto;
  max-width: 90%;   
  margin-left: 0;
  transform: translate(-50%, -60%);
  overflow-wrap: break-word;
  word-break: normal;
  white-space: pre-wrap;
  text-overflow: ellipsis;
  transition: all .35s ease-in-out;
  cursor: pointer;
}

.kgs-image_carousel .carousel:hover .slide .legend {
    background-color: rgba(51, 51, 51, 0.85);
}

.kgs-image_carousel .carousel .slide .kgs-image_carousel-icon {
  display: none;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: scale(3) translate(-10%, -35%);
  color: white;
  opacity: 0.3;
  cursor: pointer;
}

.kgs-image_carousel .carousel .slide.selected .kgs-image_carousel-icon {
  display: block;
}

@keyframes kgs-image_carousel-icon {
  0% {
      opacity: 0.5
  }
  24% {
      opacity: 0.5
  }
  25% {
      opacity: 0.3
  }
  50% {
      opacity: 0.3
  }
  75% {
      opacity: 0.3
  }
  76% {
      opacity: 0.5
  }
  100% {
      opacity: 0.5
  }
}

.kgs-image_carousel .carousel:hover .slide.selected .kgs-image_carousel-icon,
.kgs-image_carousel .carousel .slide.selected:hover .kgs-image_carousel-icon,
.kgs-image_carousel .carousel .slide.selected.legend .legend:hover + .kgs-image_carousel-icon {
  opacity: 0.5;
}


.kgs-image_carousel .carousel:hover .slide.selected .kgs-image_carousel-icon.is-animated,
.kgs-image_carousel .carousel .slide.selected:hover .kgs-image_carousel-icon.is-animated,
.kgs-image_carousel .carousel .slide.selected.legend .legend:hover + .kgs-image_carousel-icon.is-animated {
  animation: kgs-image_carousel-icon 1s infinite ease-in-out;
}
