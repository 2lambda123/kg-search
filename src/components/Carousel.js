/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

import React from "react";
import PropTypes from "prop-types";
import { isMobile } from "../helpers/BrowserHelpers";
import "./Carousel.css";


const CarouselItem = ({item, showPrevious, onPrevious, onClose, itemComponent, navigationComponent, isPreviewInstance}) => {
  const ItemComponent =  itemComponent;
  const NavigationComponent = navigationComponent;
  return (
    <div className={`kgs-carousel__item position${item.position}`} >
      <div className="kgs-carousel__content">
        <div className="kgs-carousel__header">
          {item.isActive && showPrevious && (
            <button className="kgs-carousel__previous-button" onClick={onPrevious}>
              <i className="fa fa-chevron-left" /> Previous
            </button>
          )}
          <div className="kgs-carousel__navigation">
            {item.isActive && item.data && NavigationComponent && (
              <NavigationComponent/>
            )}
          </div>
          {item.isActive && !isPreviewInstance && (
            <button className="kgs-carousel__close-button" onClick={onClose}>
              <i className="fa fa-close" />
            </button>
          )}
        </div>
        <div className="kgs-carousel__body">
          {item.isActive && item.data && ItemComponent && (
            <ItemComponent data={item.data} />
          )}
        </div>
      </div>
    </div>
  );
};

const nbOfItems = 5;

export class Carousel extends React.Component {
  constructor(props) {
    super(props);
    this.items =  Array.from(Array(nbOfItems)).map((x, idx) => ({
      id: idx,
      position: idx,
      isActive: false,
      data: null
    }));
  }
  componentDidMount() {
    if (!isMobile) {
      window.addEventListener("keyup", this._keyupHandler.bind(this), false);
    }
  }
  componentWillUnmount() {
    if (!isMobile) {
      window.removeEventListener("keyup", this._keyupHandler);
    }
  }
  _keyupHandler(event) {
    const {onPrevious, onClose, isPreviewInstance} = this.props;
    if (this.props.show) {
      if (event.keyCode === 27 && !isPreviewInstance) {
        event.preventDefault();
        typeof onClose === "function" && onClose();
      } else if (event.keyCode === 8) {
        event.preventDefault();
        typeof onPrevious === "function" && onPrevious();
      }
    }
  }
  render(){
    const {className, show, data, onPrevious, onClose, itemComponent, navigationComponent, isPreviewInstance} = this.props;
    if (!show || !Array.isArray(data) || !data.length || !itemComponent) {
      return null;
    }

    const currentPosition = (data.length -1) % nbOfItems;
    const items = this.items;
    items.forEach((item, idx) => {
      item.isActive = item.id === currentPosition;
      const position = (idx <= currentPosition?(nbOfItems - (currentPosition - idx)):(idx - currentPosition)) % 5;
      item.position = position;
      const idxData = data.length -1 - (idx <= currentPosition?(currentPosition - idx):(nbOfItems  - (idx - currentPosition)));
      item.data = idxData >= 0?data[idxData]:null;
    });
    const showPrevious = data.length > 1;

    const classNames = ["kgs-carousel", className].join(" ");

    return(
      <div className={classNames}>
        <div className="kgs-carousel__panel">
          {items.map(item => (
            <CarouselItem key={item.id} item={item} showPrevious={showPrevious} onPrevious={onPrevious} onClose={onClose} itemComponent={itemComponent} navigationComponent={navigationComponent} isPreviewInstance={isPreviewInstance} />
          ))}
        </div>
      </div>
    );
  }
}

Carousel.propTypes = {
  className: PropTypes.string,
  show: PropTypes.bool,
  data:  PropTypes.arrayOf(PropTypes.any),
  onPrevious: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  itemComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func
  ]).isRequired,
  navigationComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func
  ])
};

export default Carousel;