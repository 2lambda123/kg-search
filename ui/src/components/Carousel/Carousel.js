/*
 * Copyright 2018 - 2021 Swiss Federal Institute of Technology Lausanne (EPFL)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This open source software code was developed in part or in whole in the
 * Human Brain Project, funded from the European Union's Horizon 2020
 * Framework Programme for Research and Innovation under
 * Specific Grant Agreements No. 720270, No. 785907, and No. 945539
 * (Human Brain Project SGA1, SGA2 and SGA3).
 *
 */

import React, { useEffect, useRef, useState } from "react";
import PropTypes from "prop-types";
import { isMobile } from "../../helpers/BrowserHelpers";
import "./Carousel.css";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const getNavigation = (item, showPrevious, onClose, onPrevious, navigationComponent) => {
  const NavigationComponent = navigationComponent;
  const Navigation = () => (
    <div className="kgs-carousel__header">
      {item.isActive && showPrevious && (
        <button className="kgs-carousel__previous-button" onClick={onPrevious}>
          <FontAwesomeIcon icon="chevron-left" /> Previous
        </button>
      )}
      <div className="kgs-carousel__navigation">
        {item.isActive && item.data && NavigationComponent && (
          <NavigationComponent />
        )}
      </div>
      {item.isActive && (
        <button className="kgs-carousel__close-button" onClick={onClose}>
          <FontAwesomeIcon icon="times" />
        </button>
      )}
    </div>
  );
  Navigation.displayName = "Navigation";
  return Navigation;
};

const CarouselItem = ({ item, showPrevious, onClose, onPrevious, itemComponent, navigationComponent }) => {
  const ItemComponent = itemComponent;
  const NavigationComponent = getNavigation(item, showPrevious, onClose, onPrevious, navigationComponent);
  return (
    <div className={`kgs-carousel__item position${item.position}`}>
      <div className="kgs-carousel__content">
        {item.isActive && item.data && ItemComponent && (
          <ItemComponent data={item.data} NavigationComponent={NavigationComponent} />
        )}
      </div>
    </div>
  );
};

const nbOfItems = 5;

export const Carousel = ({ show, className, data, onPrevious, onClose, itemComponent, navigationComponent }) => {
  const [items, setItems] = useState(Array.from(Array(nbOfItems)).map((x, idx) => ({
    id: idx,
    position: idx,
    isActive: false,
    data: null
  })));

  const wrapperRef = useRef();

  useEffect(() => {
    const currentPosition = (data.length - 1) % nbOfItems;
    const result = items.map((item, idx) => {
      const idxData = data.length - 1 - (idx <= currentPosition ? (currentPosition - idx) : (nbOfItems - (idx - currentPosition)));
      const position = (idx <= currentPosition ? (nbOfItems - (currentPosition - idx)) : (idx - currentPosition)) % 5;
      return {
        isActive: item.id === currentPosition,
        position: position,
        data: idxData >= 0 ? data[idxData] : null
      };
    });
    setItems(result);
  }, []);

  useEffect(() => {
    window.instanceTabSelection = {};
    if (!isMobile) {
      const _keyupHandler = event => {
        if (show) {
          if (event.keyCode === 27) {
            event.preventDefault();
            onClose();
          }
        }
      };
      window.addEventListener("keyup", _keyupHandler, false);
      return () => window.removeEventListener("keyup", _keyupHandler);
    }
  }, []);

  if (!show || !Array.isArray(data) || !data.length || !itemComponent) {
    return null;
  }

  const handleOnClose = e => {
    if (this.wrapperRef && !this.wrapperRef.contains(e.target)) {
      window.instanceTabSelection = {};
      onClose();
    }
  };

  const showPrevious = data.length > 1;

  const classNames = ["kgs-carousel", className].join(" ");

  return (
    <div className={classNames} onClick={this.onClose}>
      <div className="kgs-carousel__panel" ref={wrapperRef}>
        {items.map(item => (
          <CarouselItem key={item.id} item={item} showPrevious={showPrevious} onPrevious={onPrevious} onClose={handleOnClose} itemComponent={itemComponent} navigationComponent={navigationComponent} />
        ))}
      </div>
    </div>
  );

};

Carousel.propTypes = {
  className: PropTypes.string,
  show: PropTypes.bool,
  data: PropTypes.arrayOf(PropTypes.any),
  onPrevious: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  itemComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
    PropTypes.object
  ]).isRequired,
  navigationComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
    PropTypes.object
  ])
};

export default Carousel;