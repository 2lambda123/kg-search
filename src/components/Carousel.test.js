/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

import React from "react";
import renderer from "react-test-renderer";
import Enzyme, { mount, shallow, render } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Carousel from "./Carousel";

Enzyme.configure({ adapter: new Adapter() });

test('Carousel component renders initially', () => {
    const component = renderer.create(
        <Carousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
  
    expect(component.toJSON()).toMatchSnapshot();
});

test('Carousel test show false"', () => {
    const component = renderer.create(
        <Carousel className="className" show={false} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    expect(component.toJSON()).toBe(null);
});
  
test('Carousel test className"', () => {
    const component = shallow(
        <Carousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    expect(component.hasClass("className"));
});

test('Carousel test number of items', () => {
    const component = render(
        <Carousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    expect(component.find(".kgs-carousel__item").length).toBe(5);
});

test('Carousel test items', () => {
    const component = renderer.create(
        <Carousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    const instance = component.getInstance();

    expect(instance.items.length).toBe(5);
    expect(instance.items[1].position).toBe(0);
    expect(instance.items[1].isActive);
    expect(instance.items[1].data.label).toBe("another label");
});

test('Carousel test navigation buttons', () => {
    const previousFn = jest.fn();
    const closeFn = jest.fn();
    const component = mount(
        <Carousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={previousFn} onClose={closeFn} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    expect(component.find(".kgs-carousel__previous-button").length).toBe(1);
    expect(component.find(".kgs-carousel__close-button").length).toBe(1);
    component.find('.kgs-carousel__previous-button').simulate('click');
    expect(previousFn.mock.calls.length).toBe(1);
    expect(closeFn.mock.calls.length).toBe(0);
    component.find('.kgs-carousel__close-button').simulate('click');
    expect(previousFn.mock.calls.length).toBe(1);
    expect(closeFn.mock.calls.length).toBe(1);
});
