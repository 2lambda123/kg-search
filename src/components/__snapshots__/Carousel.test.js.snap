// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component renders initially 1`] = `
<div
  className="kgs-carousel className"
  onClick={[Function]}
>
  <div
    className="kgs-carousel__panel"
  >
    <div
      className="kgs-carousel__item position4"
    >
      <div
        className="kgs-carousel__content"
      >
        <div
          className="kgs-carousel__header"
        >
          <div
            className="kgs-carousel__navigation"
          />
        </div>
        <div
          className="kgs-carousel__body"
        />
      </div>
    </div>
    <div
      className="kgs-carousel__item position0"
    >
      <div
        className="kgs-carousel__content"
      >
        <div
          className="kgs-carousel__header"
        >
          <button
            className="kgs-carousel__previous-button"
            onClick={[Function]}
          >
            <i
              className="fa fa-chevron-left"
            />
             Previous
          </button>
          <div
            className="kgs-carousel__navigation"
          />
          <button
            className="kgs-carousel__close-button"
            onClick={[Function]}
          >
            <i
              className="fa fa-close"
            />
          </button>
        </div>
        <div
          className="kgs-carousel__body"
        />
      </div>
    </div>
    <div
      className="kgs-carousel__item position1"
    >
      <div
        className="kgs-carousel__content"
      >
        <div
          className="kgs-carousel__header"
        >
          <div
            className="kgs-carousel__navigation"
          />
        </div>
        <div
          className="kgs-carousel__body"
        />
      </div>
    </div>
    <div
      className="kgs-carousel__item position2"
    >
      <div
        className="kgs-carousel__content"
      >
        <div
          className="kgs-carousel__header"
        >
          <div
            className="kgs-carousel__navigation"
          />
        </div>
        <div
          className="kgs-carousel__body"
        />
      </div>
    </div>
    <div
      className="kgs-carousel__item position3"
    >
      <div
        className="kgs-carousel__content"
      >
        <div
          className="kgs-carousel__header"
        >
          <div
            className="kgs-carousel__navigation"
          />
        </div>
        <div
          className="kgs-carousel__body"
        />
      </div>
    </div>
  </div>
</div>
`;
